src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs

```diff
@@ -1,82 +1,126 @@
+                JsonTokenType.None => null,
+                _ => throw new InvalidOperationException($"Unexpected token type: {reader.TokenType}")
+            };
+
+        private static List<object> ReadArray(ref Utf8JsonReader reader)
         {
-            foreach (var item in value.EnumerateArray())
+            if (reader.TokenType != JsonTokenType.StartArray)
+                throw new JsonException();
+
+            var result = new List<object>();
+
+            while (reader.Read())
             {
-                yield return ReadValue(item);
+                if (reader.TokenType == JsonTokenType.EndArray)
+                    break;
+
+                result.Add(ReadValue(ref reader));
             }
-        }
 
-        private object ReadValue(JsonElement value)
-            => value.ValueKind switch
-            {
-                JsonValueKind.Array => ReadArray(value).ToList(),
-                JsonValueKind.Object => ReadDictionary(value),
-                JsonValueKind.Number => ReadNumber(value),
-                JsonValueKind.True => true,
-                JsonValueKind.False => false,
-                JsonValueKind.String => value.GetString(),
-                JsonValueKind.Null => null,
-                JsonValueKind.Undefined => null,
-                _ => throw new InvalidOperationException($"Unexpected value kind: {value.ValueKind}")
-            };
+            return result;
+        }
 
-        private object ReadNumber(JsonElement value)
+        private static object ReadNumber(ref Utf8JsonReader reader)
         {
-            if (value.TryGetInt32(out var i))
+            if (reader.TryGetInt32(out int i))
                 return i;
-            else if (value.TryGetInt64(out var l))
+            else if (reader.TryGetInt64(out long l))
                 return l;
-            else if (BigInteger.TryParse(value.GetRawText(), out var bi))
-                return bi;
-            else if (value.TryGetDouble(out var d))
+            else if (reader.TryGetDouble(out double d))
+            {
+                // the idea is to see if there is a loss of accuracy of value
+
+                // example:
+                // value from json       : 636474637870330463636474637870330463636474637870330463
+                // successfully parsed d : 6.3647463787033043E+53
+                // new BigInteger(d)     : 636474637870330432588044308848152942336795574685138944
+                if (TryGetBigInteger(ref reader, out var bi))
+                {
+                    if (bi != new BigInteger(d))
+                        return bi;
+                }
                 return d;
-            else if (value.TryGetDecimal(out var dd))
-                return dd;
+            }
+            else if (reader.TryGetDecimal(out decimal dm))
+            {
+                // the idea is to see if there is a loss of accuracy of value
+                if (TryGetBigInteger(ref reader, out var bi))
```

Comment:
You could query if there's a converter for Big Integer from options(GetConverter) and skip the big integer if no converter present.

Reply:
Ok, but I do not want to skip.

Reply:
But at least I saw that I wrote 2 identical pieces of code.